### PR TITLE
New Technique: Access Virtual Machine using Bastion shareable link

### PR DESCRIPTION
**Pull request by siigil**
### What does this PR do?

New attack technique: Persistence via Azure Bastion shareable link.

### Motivation

This technique has been documented as a method for maintaining VM access off-network:
- https://blog.karims.cloud/2022/11/26/yet-another-azure-vm-persistence.html
- https://microsoft.github.io/Azure-Threat-Research-Matrix/Persistence/AZT509/AZT509/
- General feature info: https://learn.microsoft.com/en-us/azure/bastion/shareable-link

Currently, I'm making the assumption that shareable links are enabled (as they may already be for a contractor or similar), to focus on the creation of a shareable link as the overall technique. Open to critique on this approach!

### Checklist

- [x] The attack technique emulates a single attack step, not a full attack chain
- [x] We have factual evidence & references that the attack technique was used by real malware, pentesters, or attackers*
Researcher(s), in this case.
- [x] The attack technique makes no assumption about the state of the environment prior to warming it up

### CODE REVIEW

File: v2/internal/attacktechniques/azure/persistence/bastion-shareable-link/main.tf [hunk 1]
```diff
@@ -0,0 +1,171 @@
...
+    sku       = "2016-Datacenter"
+    version   = "latest"
+  }
+}
+
+# # # # # # # # # # # # # # # # # # # # # # # # # # # # #
+# Bastion Resource
+# # # # # # # # # # # # # # # # # # # # # # # # # # # # #
+
+# Note: Creation/destruction of a Bastion can take 10 minutes each, see https://learn.microsoft.com/en-us/azure/bastion/tutorial-create-host-portal
```
Comment (christophetd) [inline v2/internal/attacktechniques/azure/persistence/bastion-shareable-link/main.tf:125]: oh dear

File: v2/internal/attacktechniques/azure/persistence/bastion-shareable-link/main.tf [hunk 2]
```diff
@@ -0,0 +1,171 @@
...
+resource "azurerm_bastion_host" "bastion" {
+  name                = "${local.resource_prefix}-bastion-${random_string.lab_name.result}"
+  location            = azurerm_resource_group.lab_environment.location
+  resource_group_name = azurerm_resource_group.lab_environment.name
+  # Required for shareable link feature
+  sku                    = "Standard"
+  shareable_link_enabled = "true"
```
Comment (christophetd) [inline v2/internal/attacktechniques/azure/persistence/bastion-shareable-link/main.tf:132]:
```suggestion
  shareable_link_enabled = true
```

File: v2/internal/attacktechniques/azure/persistence/bastion-shareable-link/main.go [hunk 3]
```diff
@@ -0,0 +1,182 @@
...
+	"fmt"
+)
+
+//go:embed main.tf
+var tf []byte
+
+func init() {
+	const codeBlock = "```"
+	stratus.GetRegistry().RegisterAttackTechnique(&stratus.AttackTechnique{
+		ID:           "azure.persistence.bastion-shareable-link",
```
Comment (christophetd) [inline v2/internal/attacktechniques/azure/persistence/bastion-shareable-link/main.go:20]: something more action oriented like the below might work better?

```suggestion
		ID:           "azure.persistence.create-bastion-sharing-link",
```

File: v2/internal/attacktechniques/azure/persistence/bastion-shareable-link/main.go [hunk 4]
```diff
@@ -0,0 +1,182 @@
...
+
+	// Create Bastion shareable link
+	// Reference method: https://learn.microsoft.com/en-us/rest/api/virtualnetwork/put-bastion-shareable-link/put-bastion-shareable-link
+	log.Println("Getting Bastion shareable link for VM " +vmName)
+
+	poller, err := client.NewManagementClient().BeginPutBastionShareableLink(ctx, resourceGroup, bastionName, armnetwork.BastionShareableLinkListRequest{
+		VMs: []*armnetwork.BastionShareableLink{
+			{
+				VM: &armnetwork.VM{
+					ID: to.Ptr(vmId),
```
Comment (christophetd) [inline v2/internal/attacktechniques/azure/persistence/bastion-shareable-link/main.go:118]:
```suggestion
					ID: &vmId,
```

should be enough?

File: v2/internal/attacktechniques/azure/persistence/bastion-shareable-link/main.go [hunk 5]
```diff
@@ -0,0 +1,182 @@
...
+				},
+			},
+		},
+	}, nil)
+	if err != nil {
+		log.Fatalf("failed to create shareable link: %v", err)
```
Comment (christophetd) [inline v2/internal/attacktechniques/azure/persistence/bastion-shareable-link/main.go:124]: we generally try to return errors here (i.e. `return fmt.Errorf("failed to create shareable link: %v", err);`)

Comment (christophetd) [inline v2/internal/attacktechniques/azure/persistence/bastion-shareable-link/main.go:124]:
```suggestion
		return fmt.Errorf("failed to create shareable link: %v", err)
```

File: v2/internal/attacktechniques/azure/persistence/bastion-shareable-link/main.go [hunk 6]
```diff
@@ -0,0 +1,182 @@
...
+	}
+
+	_, err = poller.PollUntilDone(ctx, nil)
+	if err != nil {
+		log.Fatalf("failed to poll results: %v", err)
```
Comment (christophetd) [inline v2/internal/attacktechniques/azure/persistence/bastion-shareable-link/main.go:129]:
```suggestion
		return fmt.Errorf("failed to retrieve shareable link: %v", err)
```

File: v2/internal/attacktechniques/azure/persistence/bastion-shareable-link/main.go [hunk 7]
```diff
@@ -0,0 +1,182 @@
...
+	}
+	log.Println("Shareable link created")
+
+	// Provide URL to access Bastion shareable link
+	// NOTE: Response via Go SDK methods does not return any page contents, so we'll supply a Portal URL to fetch the link for now. (The example cited in reference link above is not clear on how to resolve this.)
+	url := fmt.Sprintln("https://portal.azure.com/#@" + tenantId + "/resource/subscriptions/" + subscriptionID + "/resourceGroups/" + resourceGroup + "/providers/Microsoft.Network/bastionHosts/" + bastionName + "/shareablelinks")
```
Comment (christophetd) [inline v2/internal/attacktechniques/azure/persistence/bastion-shareable-link/main.go:135]: sounds like Sprintf would be a bit cleaner to build this with `%s` ?

File: v2/internal/attacktechniques/azure/persistence/bastion-shareable-link/main.go [hunk 8]
```diff
@@ -0,0 +1,182 @@
...
+		VMs: []*armnetwork.BastionShareableLink{
+			{
+				VM: &armnetwork.VM{
+					ID: to.Ptr(vmId),
+				},
+			},
+		},
+	}, nil)
+	if err != nil {
+		log.Fatalf("failed to finish the request: %v", err)
```
Comment (christophetd) [inline v2/internal/attacktechniques/azure/persistence/bastion-shareable-link/main.go:172]:
```suggestion
		return fmt.Errorf("failed to delete shareable bastion link: %v", err)
```

File: v2/internal/attacktechniques/azure/persistence/bastion-shareable-link/main.go [hunk 9]
```diff
@@ -0,0 +1,182 @@
...
+	}
+	_, err = poller.PollUntilDone(ctx, nil)
+	if err != nil {
+		log.Fatalf("failed to pull the result: %v", err)
```
Comment (christophetd) [inline v2/internal/attacktechniques/azure/persistence/bastion-shareable-link/main.go:176]:
```suggestion
		return fmt.Errorf("failed to retrieve shareable bastion link deletion result: %v", err)
```

File: v2/internal/attacktechniques/azure/persistence/bastion-shareable-link/main.tf (hunk 1 above)
Comment (siigil) [reply v2/internal/attacktechniques/azure/persistence/bastion-shareable-link/main.tf:125]: Yeah, I wish there were a way to speed this up.

**Comment by siigil** [comment]
@christophetd Added changes based on your feedback in, but as a heads up haven't retested just yet (will tomorrow AM). Feel free to change/comment anything tomorrow if you get to this first!

File: v2/internal/attacktechniques/azure/persistence/bastion-shareable-link/main.go (hunk 3 above)
Comment (siigil) [reply v2/internal/attacktechniques/azure/persistence/bastion-shareable-link/main.go]: Keeping "shareable" if we can as this is the feature name, but can use "sharing" if we're up against a character limit.

Comment (christophetd) [reply v2/internal/attacktechniques/azure/persistence/bastion-shareable-link/main.go]: ah makes sense, so wdyt of `azure.persistence.create-bastion-shareable-link`?

File: v2/internal/attacktechniques/azure/persistence/create-bastion-shareable-link/main.go [hunk 10]
```diff
@@ -0,0 +1,181 @@
...
+	tenantId := params["tenant_id"]
+
+	ctx := context.Background()
+	cred := providers.Azure().GetCredentials()
+	subscriptionID := providers.Azure().SubscriptionID
+	clientOptions := providers.Azure().ClientOptions
+
+	client, err := armnetwork.NewClientFactory(subscriptionID, cred, clientOptions)
+	if err != nil {
+		log.Fatalf("failed to create client: %v", err)
```
Comment (christophetd) [inline v2/internal/attacktechniques/azure/persistence/create-bastion-shareable-link/main.go:106]:
```suggestion
		return fmt.Errorf("failed to create client: %v", err)
```

File: v2/internal/attacktechniques/azure/persistence/create-bastion-shareable-link/main.go [hunk 11]
```diff
@@ -0,0 +1,181 @@
...
+	vmName := params["vm_name"]
+
+	ctx := context.Background()
+	cred := providers.Azure().GetCredentials()
+	subscriptionID := providers.Azure().SubscriptionID
+	clientOptions := providers.Azure().ClientOptions
+
+	client, err := armnetwork.NewClientFactory(subscriptionID, cred, clientOptions)
+	if err != nil {
+		log.Fatalf("failed to create client: %v", err)
```
Comment (christophetd) [inline v2/internal/attacktechniques/azure/persistence/create-bastion-shareable-link/main.go:155]:
```suggestion
		return fmt.Errorf("failed to instantiate ARM Network client: %v", err)
```

**Comment by siigil** [comment]
Updated technique to provide shareable link + credentials for direct bastion access.

File: v2/internal/attacktechniques/azure/persistence/create-bastion-shareable-link/main.go [hunk 12]
```diff
@@ -0,0 +1,201 @@
...
+					ID: &vmId,
+				},
+			},
+		},
+	}, nil)
+
+	for pager.More() {
+		page, err := pager.NextPage(ctx)
+		if err != nil {
+			fmt.Errorf("failed to get results page: %v", err)
```
Comment (christophetd) [inline v2/internal/attacktechniques/azure/persistence/create-bastion-shareable-link/main.go:146]:
```suggestion
			return fmt.Errorf("failed to get results page: %v", err)
```

File: v2/internal/attacktechniques/azure/persistence/create-bastion-shareable-link/main.go [hunk 13]
```diff
@@ -0,0 +1,201 @@
...
+		}
+	}
+
+	log.Println("Bastion username: " + adminUsername)
+
+	// Password needs to be unmarshaled, as the resulting string from Terraform has json.HTMLEscape applied. Unmarshal is the correct operation, but string needs to be correctly formatted to work (see above).
+	var adminPasswordDecoded string
+	err = json.Unmarshal([]byte(adminPassword), &adminPasswordDecoded)
+	if err != nil{
+		fmt.Errorf("failed to unmarshal password string: %v", err)
```
Comment (christophetd) [inline v2/internal/attacktechniques/azure/persistence/create-bastion-shareable-link/main.go:159]:
```suggestion
		return fmt.Errorf("failed to unmarshal password string: %v", err)
```

**Review by christophetd — APPROVED** [review]

**Pull request merged**